src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +22,196 @@ namespace System
+                    }
+                }
+            }
+
+            return span.Slice(start, end - start + 1);
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+            => TrimHelper(ref span, ref trimElements, true, true);
+
+        /// <summary>
+        /// Removes all leading occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+            => TrimHelper(ref span, ref trimElements, true, false);
+
+        /// <summary>
+        /// Removes all trailing occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the elements are removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+            => TrimHelper(ref span, ref trimElements, false, true);
+
+        /// <summary>
+        /// Conditionally removes all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <param name="trimStart">If set to True, trims the start.</param>
+        /// <param name="trimEnd">If set to True, trims the end.</param>
+        private static ReadOnlySpan<T> TrimHelper<T>(ref ReadOnlySpan<T> span, ref ReadOnlySpan<T> trimElements, bool trimStart, bool trimEnd)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty || trimElements.IsEmpty)
```

Comment:
This guard is not present in the related `char` and `byte` overloads. Should it be?